base/coreir.jl

```diff
@@ -14,61 +14,75 @@ Core.Const
+If `field[i]` is of type `Union{}`, it means the `i`-th field is never be initialized and
+thus never be defined. In this case, `undefs[i]` should always be `true`.
+
+The same applies if `typ` is a `Tuple`, and because of how `Tuple` elements are initialized,
+`undefs[i] === false` holds except that `undefs[end]` may be `nothing` when the last element
+is `Vararg`.
 """
 Core.PartialStruct
 
-function Core.PartialStruct(typ::Type, undef::BitVector, fields::Vector{Any})
-    @assert length(undef) == length(fields) - isvarargtype(fields[end])
+function Core.PartialStruct(typ::Type, undefs::Vector{Union{Nothing,Bool}}, fields::Vector{Any})
+    fldcnt = fieldcount_noerror(typ)
+    if fldcnt !== nothing
+        @assert fldcnt == length(fields)
+    else
+        @assert typ <: Tuple && isvarargtype(fields[end])
+        @assert datatype_min_ninitialized(typ) == length(fields) - 1
+        @assert undefs[end] === nothing
+    end
+    @assert length(fields) == length(undefs)
     for i = 1:length(fields)
-        @assert fields[i] !== Union{} # TODO remove me once we start to exploit strict undef-ness of fields
+        if fields[i] === Union{}
+            @assert undefs[i] === true "`Union{}` typed field should be strictly undefined"
+        end
     end
-    return Core._PartialStruct(typ, undef, fields)
+    return Core._PartialStruct(typ, undefs, fields)
 end
 
+# Legacy constructor
 function Core.PartialStruct(@nospecialize(typ), fields::Vector{Any})
-    return Core.PartialStruct(typ, partialstruct_init_undef(typ, fields), fields)
-end
-
-partialstruct_undef_length(fields) = length(fields) - isvarargtype(fields[end])
-
-function partialstruct_init_undef(@nospecialize(typ), fields; all_defined = true)
-    n = partialstruct_undef_length(fields)
-    return partialstruct_init_undef(typ, n; all_defined)
+    return Core.PartialStruct(typ, partialstruct_init_undefs(typ, fields), fields)
 end
 
-function partialstruct_init_undef(@nospecialize(typ), n::Integer; all_defined = true)
-    all_defined && return falses(n)
-    undef = trues(n)
+partialstruct_init_undefs(@nospecialize(typ), fields::Vector{Any}) = partialstruct_init_undefs(typ, length(fields))
+function partialstruct_init_undefs(@nospecialize(typ), n::Int)
+    undefs = Union{Nothing,Bool}[nothing for _ in 1:n]
     for i in 1:min(datatype_min_ninitialized(typ), n)
-        undef[i] = false
+        undefs[i] = false
     end
-    return undef
+    return undefs
 end
 
-(==)(a::PartialStruct, b::PartialStruct) = a.typ === b.typ && a.undef == b.undef && a.fields == b.fields
+a::PartialStruct == b::PartialStruct = a.typ === b.typ && a.undefs == b.undefs && a.fields == b.fields
 
 function Base.getproperty(pstruct::Core.PartialStruct, name::Symbol)
-    name === :undef && return getfield(pstruct, :undef)::BitVector
+    name === :undefs && return getfield(pstruct, :undefs)::Vector{Union{Nothing,Bool}}
```

Comment:
NB: this method is not currently used from Compiler.jl. PR #57553 aims to fix that.